server/src/main/java/org/opensearch/common/util/concurrent/QueueResizableOpenSearchThreadPoolExecutor.java

```diff
@@ -0,0 +1,119 @@
+        int maximumPoolSize,
+        long keepAliveTime,
+        TimeUnit unit,
+        ResizableBlockingQueue<Runnable> workQueue,
+        Function<Runnable, WrappedRunnable> runnableWrapper,
+        ThreadFactory threadFactory,
+        XRejectedExecutionHandler handler,
+        ThreadContext contextHolder
+    ) {
+        super(name, corePoolSize, maximumPoolSize, keepAliveTime, unit, workQueue, threadFactory, handler, contextHolder);
+        this.workQueue = workQueue;
+        this.runnableWrapper = runnableWrapper;
+        this.executionEWMA = new ExponentiallyWeightedMovingAverage(EWMA_ALPHA, 0);
+    }
+
+    @Override
+    protected Runnable wrapRunnable(Runnable command) {
+        return super.wrapRunnable(this.runnableWrapper.apply(command));
+    }
+
+    @Override
+    protected Runnable unwrap(Runnable runnable) {
+        final Runnable unwrapped = super.unwrap(runnable);
+        if (unwrapped instanceof WrappedRunnable) {
+            return ((WrappedRunnable) unwrapped).unwrap();
+        } else {
+            return unwrapped;
+        }
+    }
+
+    /**
+     * Returns the exponentially weighted moving average of the task execution time
+     */
+    @Override
+    public double getTaskExecutionEWMA() {
+        return executionEWMA.getAverage();
+    }
+
+    /**
+     * Returns the current queue size (operations that are queued)
+     */
+    @Override
+    public int getCurrentQueueSize() {
+        return workQueue.size();
+    }
+
+    @Override
+    protected void afterExecute(Runnable r, Throwable t) {
+        super.afterExecute(r, t);
+        // A task has been completed, it has left the building. We should now be able to get the
+        // total time as a combination of the time in the queue and time spent running the task. We
+        // only want runnables that did not throw errors though, because they could be fast-failures
+        // that throw off our timings, so only check when t is null.
+        assert super.unwrap(r) instanceof TimedRunnable : "expected only TimedRunnables in queue";
+        final TimedRunnable timedRunnable = (TimedRunnable) super.unwrap(r);
+        final boolean failedOrRejected = timedRunnable.getFailedOrRejected();
+
+        final long taskExecutionNanos = timedRunnable.getTotalExecutionNanos();
+        assert taskExecutionNanos >= 0 || (failedOrRejected && taskExecutionNanos == -1)
+            : "expected task to always take longer than 0 nanoseconds or have '-1' failure code, got: "
+                + taskExecutionNanos
+                + ", failedOrRejected: "
+                + failedOrRejected;
+
+        if (taskExecutionNanos != -1) {
+            // taskExecutionNanos may be -1 if the task threw an exception
+            executionEWMA.addValue(taskExecutionNanos);
+        }
+    }
+
+    /**
+     * Resizes the work queue capacity of the pool
+     * @param capacity the new capacity
+     */
+    public synchronized int resize(int capacity) {
```

Comment:
For my understanding who calls resize?

Reply:
AFAIK that's could be done from the plugin(s), as per attached issue

Reply:
Should the plugin have the capability able to override the resize logic? Do you think we could expose a contract?

Reply:
I believe this is the whole purpose of the issue and the change (I am finalizing the https://github.com/opensearch-project/OpenSearch/pull/826 since the pull request was abandoned).